### PR TITLE
Implement the rest of the vertex_state correctness tests.

### PR DESCRIPTION
**Pull request by Kangz**
The plan mentioned a lot of test combination with indexed=true/false but
this seems unnecessary since we already have different tests that check
the "index" generation for indexed calls.

<hr>

**Author checklist for test code/plans:**

- [X] All outstanding work is tracked with "TODO" in a test/file description or `.unimplemented()` on a test.
- [X] New helpers, if any, are documented using `/** doc comments */` and can be found via `helper_index.txt`.
- [X] (Optional, sometimes not possible.) Tests pass (or partially pass without unexpected issues) in an implementation. (Add any extra details above.)

**[Reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md) for test code/plans:** (Note: feel free to pull in other reviewers at any time for any reason.)

- [ ] The test path is reasonable, the [description](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) "describes the test, succinctly, but in enough detail that a reader can read only the test plans in a file or directory and evaluate the completeness of the test coverage."
- [ ] Tests appear to cover this area completely, except for outstanding TODOs. Validation tests use control cases.
    (This is critical for coverage. Assume anything without a TODO will be forgotten about forever.)
- [ ] Existing (or new) test helpers are used where they would reduce complexity.
- [ ] TypeScript code is readable and understandable (is unobtrusive, has reasonable type-safety/verbosity/dynamicity).

### CODE REVIEW

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 1]
```diff
@@ -789,3 +747,333 @@ g.test('buffersWithVaryingStepMode')
     }));
     t.runTest(state);
   });
+
+g.test('vertexBufferUsedMultipleTime_overlapped')
```
Comment (austinEng) [inline src/webgpu/api/operation/vertex_state/correctness.spec.ts:751]: nit: I think test names are usually camel_case unless they are literally some API construct like "GPUBufferUsage"

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 2]
```diff
@@ -789,3 +747,333 @@ g.test('buffersWithVaryingStepMode')
...
+    const pipeline = t.makeTestPipeline(testData, kVertexCount, kInstanceCount);
+    const expectedDataBG = t.createExpectedBG(testData, pipeline);
+    t.submitRenderPass(pipeline, vertexBuffers, expectedDataBG, kVertexCount, kInstanceCount);
+  });
+
+g.test('maxBuffersAndAttribs')
+  .desc(
+    `Test a vertex state that loads as many attributes and buffers as possible.
+  - For each format.
+  TODO find a way to test maxAttribs. Right now this test is gated on kMaxUniformBuffersPerStage
```
Comment (austinEng) [inline src/webgpu/api/operation/vertex_state/correctness.spec.ts:908]: Can we modify the generated shader to read from both uniform buffers and storage buffers?

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 3]
```diff
@@ -789,3 +747,333 @@ g.test('buffersWithVaryingStepMode')
...
+          kMaxVertexBufferArrayStride - formatSize - componentSize,
+          kMaxVertexBufferArrayStride - formatSize,
+        ]);
+      })
+  )
+  .fn(t => {
+    const { format, stepMode, offset } = t.params;
+    const kCount = 10;
+
+    // Create the stride 0 part of the test, first be faking a single vertex being drawn and
```
Comment (austinEng) [inline src/webgpu/api/operation/vertex_state/correctness.spec.ts:973]:
```suggestion
    // Create the stride 0 part of the test, first by faking a single vertex being drawn and
```

**Review by austinEng — APPROVED** [review]
LGTM, but I only briefly looked at existing code in `makeTestPipeline, `createExpectedBG`, `submitRenderPass`, `runTest`, etc. for context.

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts [hunk 4]
```diff
@@ -789,3 +747,333 @@ g.test('buffersWithVaryingStepMode')
...
+    const { format, vbCount, additionalVBOffset } = t.params;
+    const kVertexCount = 20;
+    const kInstanceCount = 1;
+    const formatInfo = kVertexFormatInfo[format];
+    const formatByteSize = formatInfo.bytesPerComponent * formatInfo.componentCount;
+    // We need to align so the offset for non-0 setVertexBuffer don't fail validation.
+    const alignedFormatByteSize = align(formatByteSize, 4);
+
+    // Create the data that will be in the vertex buffer. It is larger than kVertexCount because
+    // each vertex buffer will be offset by one whole vertexData (so essentially) from the previous one.
```
[the remark this answers is not in the document]
Comment (austinEng) [reply src/webgpu/api/operation/vertex_state/correctness.spec.ts:774]: 2nd pass looking at this: I think we could use more of an explanation of why we're making a `baseData` and then transforming it into the `testData` list. Reading the test, it's not directly obvious what is being tested without carefully doing the baseData => testData transformation in my head.

Same for the test after this one.

**Comment by kainino0x** [comment]
Please let me know if you want a second review, I am skipping this for now to burn down my review inbox.

File: src/webgpu/api/operation/vertex_state/correctness.spec.ts (hunk 2 above)
Comment (Kangz) [reply src/webgpu/api/operation/vertex_state/correctness.spec.ts:908]: We can. I'd like to do that in a follow-up PR if you don't mind.

Comment (austinEng) [reply src/webgpu/api/operation/vertex_state/correctness.spec.ts:908]: Sounds fine

**Comment by Kangz** [comment]
Addressed all comments.

**Comment by Kangz** [comment]
Is this ok to merge?

**Comment by austinEng** [comment]
I think so, still LGTM

**Comment by Kangz** [comment]
Yay! Thanks

**Pull request merged**